test/fail_compilation/failcstuff2.c

```diff
@@ -30,8 +30,6 @@ fail_compilation/failcstuff2.c(205): Error: variable `var` is used as a type
 fail_compilation/failcstuff2.c(203):        variable `var` is declared here
 fail_compilation/failcstuff2.c(254): Error: identifier or `(` expected before `)`
 fail_compilation/failcstuff2.c(255): Error: identifier or `(` expected
-fail_compilation/failcstuff2.c(302): Error: incompatible types for `(buf) is (1)`: `ubyte*` and `int`
-fail_compilation/failcstuff2.c(304): Error: incompatible types for `(2) is (buf)`: `int` and `ubyte*`
```

Comment:
This PR looks like a refactoring, however, it seems that it is also changing behavior.

Reply:
You're right. That is a case not handled correctly by the existing code. This refactoring does handle it correctly.

Reply:
Probably want to move that to compilable then. I think I put it here because gcc was erroring with such code, but maybe I was erroneously testing with warnings as errors.

Reply:
@WalterBright Please move the test to compilable so that we can go ahead and merge this.